pulsar-common/src/main/java/com/yahoo/pulsar/common/util/collections/ConcurrentLongPairSet.java

```diff
@@ -528,4 +529,23 @@ public int compareTo(LongPair o) {
             }
         }
     }
+    
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append('{');
+        final AtomicBoolean first = new AtomicBoolean(true);
+        forEach((item1, item2) -> {
+            if (!first.getAndSet(false)) {
+                sb.append(", ");
+            }
+            sb.append("[");
```

Comment:
does forEach() give you a consistent view of the data structure, or can it still change while you're iterating?

if it can change, you might get both better logs + better performance if you had a way to create a local consistent copy in a "plain" map and then work on that so you don't need getAndSet.